src/org/exist/indexing/IndexController.java

```diff
@@ -202,22 +208,33 @@ public void removeCollection(final Collection collection, final DBBroker broker,
     /**
      * Re-index all nodes below the specified root node, using the given mode.
      *
-     * @param transaction the current transaction
+     * @param txn the current transaction
      * @param reindexRoot the node from which reindexing should occur
      * @param mode the mode, one of {@link ReindexMode#UNKNOWN}, {@link ReindexMode#STORE},
      * {@link ReindexMode#REMOVE_SOME_NODES} or {@link ReindexMode#REMOVE_ALL_NODES}.
      */
-    public void reindex(final Txn transaction, IStoredNode<? extends IStoredNode> reindexRoot, final ReindexMode mode) {
+    public void reindex(final Txn txn, final IStoredNode<? extends IStoredNode> reindexRoot, final ReindexMode mode) {
         if (reindexRoot == null) {
             return;
         }
+        IStoredNode<? extends IStoredNode> node = broker.objectWith(new NodeProxy(reindexRoot.getOwnerDocument(), reindexRoot.getNodeId()));
+
         setReindexing(true);
-        reindexRoot = broker.objectWith(new NodeProxy(reindexRoot.getOwnerDocument(), reindexRoot.getNodeId()));
-        setDocument(reindexRoot.getOwnerDocument(), mode);
-        getStreamListener();
-        IndexUtils.scanNode(broker, transaction, reindexRoot, listener);
-        flush();
-        setReindexing(false);
+        try {
+            StreamListener stream = getStreamListener(node.getOwnerDocument(), mode);
+            try {
+                stream.startIndexDocument(txn);
+
+                IndexUtils.scanNode(broker, txn, node, listener);
+            } finally {
+                stream.endIndexDocument(txn);
+            }
+
+            //that must be called outside this method to avoid slowdown
```

Comment:
I don't think that is the case; In your changes, `NativeBroker#flush()` is still called from many of those operations, and that they calls `IndexController#flush()`, so the index is still being flushed on many operations. I think you are trying to avoid that?

Also, if I am not mistaken, the problem with moving `flush` to `TransactionManager#processSystemTasks` within eXist is that this will prohibit RYOW on the index within a transaction. Possibly also deferring all the index writes to large batches that will be single-threaded will slow down transaction concurrency.